### ISSUE TITLE
Plymouth not showing on reboot or shutdown

### ISSUE THREAD

**Issue by tobiashort** [NixOS/nixpkgs#191620]
### Describe the bug
The following I am experiencing on the current stable version (22.05) and unstable version (22.11). When having plymouth activated like seen in the code snippet below, the splash screen does not show or does only show some milliseconds when rebooting or shutting down. Instead some messages are displayed, that look like error messages to me.

```console
  # Boot splash
  boot.plymouth.enable = true;
  boot.plymouth.theme = "breeze";
  boot.initrd.systemd.enable = true;
```

```console
[    87.778393] sd-umoun[2335]: Failed to unmount /oldroot/nix/store: Device or resource busy
[    87.778853] sd-umoun[2336]: Failed to unmount /oldroot/dev: Device or resource busy
[    87.779264] sd-umoun[2337]: Failed to unmount /oldroot: Device or resource busy
[    87.779715] shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[    87.780182] shutdown[1]: Failed to finalize file systems, DM devices, ignoring.
```

Those messages I cannot find in any journal or log file. I had to record it by video.

I did some researches and I could find some posts with similar issues in other distributions. It may has something to do with full disk encryption via luks which is also enabled on my machine. In some cases it might have something to do with Nvidia, but in my case I have only an integrated Intel GPU. So I could not find a  solution.

The messages may be not that harmful and I could live with ignoring them, but I still want to see my splash screen on shutdown or reboot, which is not the case.

### Steps To Reproduce
1. Enable boot splash as described above
2. Shutdown or reboot the system

### Expected behavior
The splash screen is showing when rebooting or shutting down, instead of those messages.

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.67, NixOS, 22.11 (Raccoon), 22.11pre408827.1158501e7c7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

I did switch to unstable to test this behaviour there

**Comment by phpcitizen** [comment]
Same happens to me.

**Comment by SuperSandro2000** [comment]
Is this still an issue for 23.11?

**Comment by andresilva** [comment]
This is still an issue.

**Comment by SuperSandro2000** [comment]
Please provide some logs. Are they still the same as above?

**Comment by andresilva** [comment]
Like in OP I see this message:

```
[    87.779715] shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[    87.780182] shutdown[1]: Failed to finalize file systems, DM devices, ignoring.
```

I can't find the message in `journalctl -b-1`, I recorded it on video as the computer shuts down right after it shows. plymouth does work on shutdown if theme isn't set to `breeze`, just using the default. But the error still shows before the computer shuts down.

**Comment by cccslater** [comment]
No matter what theme I set, Plymouth does not show on shutdown/reboot. It thinks it's showing, though:
```
Mar 31 12:59:09 hp systemd[1]: Starting Show Plymouth Power Off Screen...
```

Other lines mentioning Plymouth:
```
Mar 31 12:59:09 hp systemd[1]: plymouth-quit-wait.service: Deactivated successfully.
...
Mar 31 12:59:09 hp systemd[1]: Received SIGRTMIN+20 from PID 1831 (plymouthd).
...
Mar 31 12:59:09 hp systemd[1]: Received SIGRTMIN+20 from PID 1831 (plymouthd).
Mar 31 12:59:09 hp systemd[1]: Started Show Plymouth Power Off Screen.
```

I don't get the above errors about unmounting.

**Comment by andresilva** [comment]
> Do you have any filesystems mounted via eg luks that nixos isn't aware off and that isn't in the fstab?

No. And the filesystem it's complaining about is the root filesystem. I found this: https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/88. I can confirm that once I enabled plymouth then `/var/log/boot.log` started being written to.

**Comment by cccslater** [comment]
> Did you try setting none, so not setting the option at all?
> 
> What graphics system do you have?

Yes, usually I have `bgrt` and tested with removing that line (default setting). It's AMD integrated graphics on a laptop, can't give more detail than that at present as I'm on a different machine.

**Comment by andresilva** [comment]
There are multiple issues here. In my case the reason for plymouth to not show up in shutdown is because the shutdown is too fast. Adding a dummy systemd service that delays shut down made plymouth show up:

```
  systemd.services.shutdown-delay = {
    wantedBy = [ "multi-user.target" ];
    before = [ "shutdown.target" ];
    serviceConfig = {
      Type = "simple";
      ExecStart = "${pkgs.coreutils}/bin/sleep infinity";
      ExecStopPost = "${pkgs.coreutils}/bin/sleep infinity";
      TimeoutStopSec = 20;
    };
  };
```

For the issue regarding rootfs not being unmounted on shutdown: #395008.

**Comment by cccslater** [comment]
> systemd.services.shutdown-delay = {
>     wantedBy = [ "multi-user.target" ];
>     before = [ "shutdown.target" ];
>     serviceConfig = {
>       Type = "simple";
>       ExecStart = "${pkgs.coreutils}/bin/sleep infinity";
>       ExecStopPost = "${pkgs.coreutils}/bin/sleep infinity";
>       TimeoutStopSec = 20;
>     };
>   };

This doesn't fix it for me, and I'm not having a rootfs error... Some other issue!

**Comment by andresilva** [comment]
Can you test whether #395008 fixes it for you?

**Comment by SuperSandro2000** [comment]
Thanks for opening a PR. I give it a try if it breaks things for me or makes things smoother.

**Comment by alexeev-prog** [comment]
This is already opened?

**Comment by cccslater** [comment]
What do you mean by that?

**Comment by alexeev-prog** [comment]
My friend got this error. But he does not using plymouth, he just entered `systemctl shutdown` but system is not shutdown.

Sorry for my bad english please. I add his screenshot.

<img width="960" height="1280" alt="Image" src="https://github.com/user-attachments/assets/889ac856-c553-4437-ad17-dfe990cb9bf9" />
